File: PrimaMateria/PrimaMateria/Classes/XTRNuclidesIsotopesViewController.m
//
//  XTRNuclidesIsotopesViewController.m
//  PrimaMateria
//
//  Created by Jerry Porter on 3/5/2010.
//  Copyright 2016 xTrensa. All rights reserved.
//

@interface XTRNuclidesIsotopesViewController ()
- (UILabel *) tableCellLabelWithXPos: (float) xPos YPos: (float) yPos width: (float) aWidth height: (float) aHeight property: (NSObject *) aProperty columnPosition: (int) aColumnPosition modulus: (int) aModulus forCell: (DynoTableCell *) cell;
@end

@implementation XTRNuclidesIsotopesViewController

#pragma mark Private Methods

- (UILabel *) tableCellLabelWithXPos: (float) xPos YPos: (float) yPos width: (float) aWidth height: (float) aHeight property: (NSObject *) aProperty columnPosition: (int) aColumnPosition modulus: (int) aModulus forCell: (DynoTableCell *) cell {
    UILabel *label = [[UILabel alloc] initWithFrame: CGRectMake(xPos, yPos, aWidth, aHeight)];
    if (aModulus == 0)
        label.backgroundColor = UIColor.whiteColor;
    else
        label.backgroundColor = [UIColor colorWithHue: 219 / HSB_CONSTANT saturation: 0.1f brightness: 1.0f alpha: 1.0f];
    [cell addColumn: aColumnPosition];
    label.font = [UIFont systemFontOfSize:20.0];
    label.textAlignment =  NSTextAlignmentCenter;
    label.textColor = UIColor.blackColor;
    label.text = [NSString stringWithFormat: @" %@", aProperty];
    [cell.contentView addSubview: label];
    return label;
}

#pragma mark - Misc Methods

- (void)setupUI {
    if (self.element != nil) {
        if (self.tableView != nil) {
            [self.tableView removeFromSuperview];
            self.tableView = nil;
        }
        self.tableView = [[UITableView alloc] initWithFrame:self.swapView.frame style: UITableViewStylePlain];
        self.tableView.alwaysBounceVertical = NO;
        self.tableView.alwaysBounceHorizontal = NO;
        self.tableView.dataSource = self;
        self.tableView.delegate = self;
        self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
        self.tableView.backgroundColor = UIColor.blackColor;
        self.tableView.rowHeight = 34.0f;
        [self.view addSubview: self.tableView];
        self.nuclidesAndIsotopesArray = self.element.nuclidesAndIsotopes;
        [self.tableView reloadData];
    }
}

#pragma mark - UITableView DataSource Methods

- (UITableViewCell *) tableView: (UITableView *) aTableView cellForRowAtIndexPath: (NSIndexPath *) indexPath {
    NSUInteger row = indexPath.row;
    NSString *MyIdentifier = [NSString stringWithFormat: @"Column %i", row];
    DynoTableCell *cell = [aTableView dequeueReusableCellWithIdentifier: MyIdentifier];
    if (cell == nil) {
        NSInteger modulus = row % 2;
        NSDictionary *dict = self.nuclidesAndIsotopesArray[row];
        cell = [[DynoTableCell alloc] initWithStyle: UITableViewCellStyleDefault reuseIdentifier: MyIdentifier];
        [self tableCellLabelWithXPos: 0.0f   YPos: 0.0f width: 103.0f  height: 32.0f property:[dict valueForKey: @"nuclide"] columnPosition: 1 modulus: modulus forCell: cell];
        [self tableCellLabelWithXPos: 104.0f  YPos: 0.0f width: 103.0f height: 32.0f property:[dict valueForKey: ELEMENT_ATOMIC_MASS] columnPosition: 2 modulus: modulus forCell: cell];
        [self tableCellLabelWithXPos: 208.0f YPos: 0.0f width: 101.0f height: 32.0f property:[dict valueForKey: @"nn"] columnPosition: 3 modulus: modulus forCell: cell];
        [self tableCellLabelWithXPos: 310.0f YPos: 0.0f width: 101.0f height: 32.0f property:[dict valueForKey: @"abun"] columnPosition: 4 modulus: modulus forCell: cell];
        [self tableCellLabelWithXPos: 412.0f YPos: 0.0f width: 101.0f height: 32.0f property:[dict valueForKey: @"spin"] columnPosition: 5 modulus: modulus forCell: cell];
        [self tableCellLabelWithXPos: 514.0f YPos: 0.0f width: 101.0f height: 32.0f property:[dict valueForKey: @"halfLife"] columnPosition: 6 modulus: modulus forCell: cell];
        [self tableCellLabelWithXPos: 616.0f YPos: 0.0f width: 101.0f height: 32.0f property:[dict valueForKey: @"dm"] columnPosition: 7 modulus: modulus forCell: cell];
        [self tableCellLabelWithXPos: 718.0f YPos: 0.0f width: 101.0f height: 32.0f property:[dict valueForKey: @"dt"] columnPosition: 8 modulus: modulus forCell: cell];
        [self tableCellLabelWithXPos: 820.0f YPos: 0.0f width: 101.0f height: 32.0f property:[dict valueForKey: @"br"] columnPosition: 9 modulus: modulus forCell: cell];
        [self tableCellLabelWithXPos: 922.0f YPos: 0.0f width: 101.0f height: 32.0f property:[dict valueForKey: @"decayEnergy"] columnPosition: 10 modulus: modulus forCell: cell];
    }
    return cell;
}

- (NSInteger) numberOfSectionsInTableView: (UITableView *) aTableView {
    return 1;
}

- (NSInteger) tableView: (UITableView *) aTableView numberOfRowsInSection: (NSInteger) section {
    return self.element.nuclidesAndIsotopes.count;
}

#pragma mark - View Management Methods

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.swapView removeFromSuperview];
}

- (BOOL) shouldAutorotateToInterfaceOrientation: (UIInterfaceOrientation) interfaceOrientation {
    return UIInterfaceOrientationIsLandscape(interfaceOrientation);
}

#pragma mark - Memory Management Methods

- (void)dealloc {
    self.tableView.delegate = nil;
    self.tableView = nil;
    self.headerView = nil;
    self.swapView = nil;
}

@end

File: PrimaMateria/PrimaMateria/Classes/XTRNuclidesIsotopesViewController.swift
//
//  XTRNuclidesIsotopesViewController.swift
//  PrimaMateria
//
//  Created by Jerry Porter on 4/20/16.
//  Copyright ©2018 xTrensa. All rights reserved.
//

class XTRNuclidesIsotopesViewController : XTRSwapableViewController, UITableViewDelegate, UITableViewDataSource {
    
    @IBOutlet var headerView : UIView!
    @IBOutlet var swapView: UIView!
    
    private var nuclidesAndIsotopesArray : [[String : AnyObject]]?
    private var tableView : UITableView?
    
    // MARK: - Initialization Methods
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)!
    }
    
    // MARK: - Internal Methods
    
    func tableCellLabelWithXPos(_ xPos: CGFloat, yPos: CGFloat, aWidth: CGFloat, aHeight: CGFloat, property: String, columnPosition: Int, aModulus: Int, cell: XTRTableCell) {
        let label = UILabel(frame: CGRect(x: xPos, y: yPos, width: aWidth, height: aHeight))
        
        label.backgroundColor = (aModulus == 0) ? UIColor.white : XTRColorFactory.rowColor
        label.font = UIFont.systemFont(ofSize: 14.0)
        label.textAlignment = .center
        label.textColor = UIColor.black
        label.text = property
        label.adjustsFontSizeToFitWidth = true

        cell.backgroundView?.backgroundColor = UIColor.black
        cell.contentView.backgroundColor = UIColor.black
        cell.addColumn(columnPosition)
        cell.contentView.addSubview(label)
    }
    
    // MARK: - Misc Methods
    
    override func setupUI() {
        if element != nil {
            if tableView != nil {
                tableView!.removeFromSuperview()
                tableView = nil
            }
            
            tableView = UITableView(frame: swapView.frame, style: .plain)

            tableView!.alwaysBounceVertical = false
            tableView!.alwaysBounceHorizontal = false
            tableView!.dataSource = self
            tableView!.delegate = self
            tableView!.separatorStyle = .none
            tableView!.allowsSelection = false
            tableView!.backgroundColor = UIColor.black
            tableView!.backgroundView = UIView(frame: swapView.frame)
            tableView!.backgroundView!.backgroundColor = UIColor.black
            tableView!.isOpaque = false
            nuclidesAndIsotopesArray = element!.nuclidesAndIsotopes as! [[String : AnyObject]]?
            view.addSubview(tableView!)
            tableView!.reloadData()
        }
    }
    
    // MARK: - Action Methods
    
    // MARK: - View Management Methods
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        swapView.removeFromSuperview()
    }
    
    override var shouldAutorotate : Bool {
        return false
    }
    
    override var supportedInterfaceOrientations : UIInterfaceOrientationMask {
        return .landscape
    }
    
    // MARK: - Memory Management Methods
    
    deinit {
        headerView = nil
        swapView = nil
    }
    
}

extension XTRNuclidesIsotopesViewController {
    
    // MARK:- UITableView DataSource Methods
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let row = indexPath.row
        let MyIdentifier = "Column \(row)"
        
        var cell : XTRTableCell? = tableView.dequeueReusableCell(withIdentifier: MyIdentifier) as? XTRTableCell
        
        if cell == nil {
            let modulus = row % 2
            let dict = nuclidesAndIsotopesArray![Int(row)]
            let nuclideSymbol : String? = dict[ELEMENT_NUCLIDE_SYMBOL] as? String
            let zp : String? = dict[ELEMENT_NUCLIDE_ZP] as? String
            let nn : String? = dict[ELEMENT_NUCLIDE_NN] as? String
            let isotopicMass : String? = dict[ELEMENT_ATOMIC_MASS] as? String
            let halfLife : String? = dict[ELEMENT_NUCLIDE_HALF_LIFE] as? String
            let decayModes : String? = dict[ELEMENT_NUCLIDE_DECAY_MODES] as? String
            let decayEnergy : String? = dict[ELEMENT_NUCLIDE_DECAY_ENERGY] as? String
            let daughterIsotopes : String? = dict[ELEMENT_NUCLIDE_DAUGHTER_ISOTOPES] as? String
            let nuclearSpin : String? = dict[ELEMENT_NUCLIDE_NUCLEAR_SPIN] as? String
            let isotopicComposition : String? = dict[ELEMENT_NUCLIDE_ISOTOPIC_COMPOSITION] as? String
            let naturalRangeVariation : String? = dict[ELEMENT_NUCLIDE_NATURAL_RANGE_VARIATION] as? String

            cell = XTRTableCell(style: .default, reuseIdentifier: MyIdentifier)
            
            tableCellLabelWithXPos(0, yPos: 0, aWidth: 63, aHeight: 32, property: (nuclideSymbol != nil) ? nuclideSymbol! : STRING_EMPTY, columnPosition: 1, aModulus: modulus, cell: cell!)
            
            tableCellLabelWithXPos(64, yPos: 0, aWidth: 41, aHeight: 32, property: (zp != nil) ? zp! : STRING_EMPTY, columnPosition: 2, aModulus: modulus, cell: cell!)

            tableCellLabelWithXPos(106, yPos: 0, aWidth: 41, aHeight: 32, property: (nn != nil) ? nn! : STRING_EMPTY, columnPosition: 3, aModulus: modulus, cell: cell!)

            tableCellLabelWithXPos(148, yPos: 0, aWidth: 124, aHeight: 32, property: (isotopicMass != nil) ? isotopicMass! : STRING_EMPTY, columnPosition: 4, aModulus: modulus, cell: cell!)

            tableCellLabelWithXPos(273, yPos: 0, aWidth: 131, aHeight: 32, property: (halfLife != nil) ? halfLife! : STRING_EMPTY, columnPosition: 5, aModulus: modulus, cell: cell!)
            
            tableCellLabelWithXPos(405, yPos: 0, aWidth: 118, aHeight: 32, property: (decayModes != nil) ? decayModes! : STRING_EMPTY, columnPosition: 6, aModulus: modulus, cell: cell!)
            
            tableCellLabelWithXPos(524, yPos: 0, aWidth: 122, aHeight: 32, property: (decayEnergy != nil) ? decayEnergy! : STRING_EMPTY, columnPosition: 7, aModulus: modulus, cell: cell!)

            tableCellLabelWithXPos(647, yPos: 0, aWidth: 78, aHeight: 32, property: (daughterIsotopes != nil) ? daughterIsotopes! : STRING_EMPTY, columnPosition: 8, aModulus: modulus, cell: cell!)

            tableCellLabelWithXPos(726, yPos: 0, aWidth: 60, aHeight: 32, property: (nuclearSpin != nil) ? nuclearSpin! : STRING_EMPTY, columnPosition: 9, aModulus: modulus, cell: cell!)

            tableCellLabelWithXPos(787, yPos: 0, aWidth: 118, aHeight: 32, property: (isotopicComposition != nil) ? isotopicComposition! : STRING_EMPTY, columnPosition: 10, aModulus: modulus, cell: cell!)

            tableCellLabelWithXPos(906, yPos: 0, aWidth: 118, aHeight: 32, property: (naturalRangeVariation != nil) ? naturalRangeVariation! : STRING_EMPTY, columnPosition: 11, aModulus: modulus, cell: cell!)
        }
        
        return cell!
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return element!.nuclidesAndIsotopes!.count
    }
    
    func tableView(_ aTableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 34
    }

}
